dali/operators/image/crop/bbox_crop.cc

```diff
@@ -14,143 +14,650 @@
+    std::vector<BoundingBox> new_bboxes;
+    std::vector<int> new_labels;
+    bool process_labels = !labels.empty();
+    assert(labels.empty() || labels.size() == bboxes.size());
+    for (size_t i = 0; i < bboxes.size(); i++) {
+      if (crop.Contains(bboxes[i].Centroid())) {
+        new_bboxes.push_back(bboxes[i]);
+        if (process_labels)
+          new_labels.push_back(labels[i]);
+      }
+    }
+    std::swap(bboxes, new_bboxes);
+    if (process_labels)
+      std::swap(labels, new_labels);
+  }
+
+  void WriteCropToOutput(SampleWorkspace &ws, const BoundingBox &crop) {
+    const int ndim = crop.ndim();
+    const auto coordinates = crop.AsStartAndShape();
+
+    // output0 : anchor, output1 : shape
+    auto &anchor_out = ws.Output<CPUBackend>(0);
+    anchor_out.Resize({ndim});
+    auto *anchor_out_data = anchor_out.mutable_data<float>();
+
+    auto &shape_out = ws.Output<CPUBackend>(1);
+    shape_out.Resize({ndim});
+    auto *shape_out_data = shape_out.mutable_data<float>();
+
+    for (int dim = 0; dim < ndim; dim++) {
+      anchor_out_data[dim] = coordinates[dim];
+      shape_out_data[dim]  = coordinates[ndim + dim];
+    }
+  }
+
+  void WriteBoxesToOutput(SampleWorkspace &ws, const std::vector<BoundingBox> &bounding_boxes) {
+    int box_size = 2 * ndim_;
+    auto &bbox_out = ws.Output<CPUBackend>(2);
+    bbox_out.Resize({static_cast<int64_t>(bounding_boxes.size()), box_size});
+    auto *bbox_out_data = bbox_out.mutable_data<float>();
+
+    for (size_t i = 0; i < bounding_boxes.size(); ++i) {
+      auto *output = bbox_out_data + i * box_size;
+      assert(bounding_boxes[i].ndim() == ndim_);
+      auto coordinates = bounding_boxes[i].As(bbox_layout_);
+      for (int j = 0; j < box_size; j++)
+        output[j] = coordinates[j];
+    }
+  }
+
+  void WriteLabelsToOutput(SampleWorkspace &ws, const std::vector<int> &labels) {
+    auto &labels_out = ws.Output<CPUBackend>(3);
+    labels_out.Resize({static_cast<Index>(labels.size()), 1});
+    auto *labels_out_data = labels_out.mutable_data<int>();
+    for (size_t i = 0; i < labels.size(); i++) {
+      labels_out_data[i] = labels[i];
+    }
+  }
+
+ private:
+  OpSpec spec_;
+  int num_attempts_;
+  bool has_labels_;
+  bool has_crop_shape_;
+  bool has_input_shape_;
+
+  Range scale_range_;
+  Range aspect_ratio_range_;
+
+  TensorLayout bbox_layout_;
+  TensorLayout shape_layout_;
 
-  WriteCropToOutput(ws, prospective_crop.crop);
-  WriteBoxesToOutput(ws, selected_boxes);
-  WriteLabelsToOutput(ws, selected_labels);
+  BatchRNG<std::mt19937> rngs_;
+
+  std::vector<SampleOption> sample_options_;
+
+  std::vector<TensorShape<>> crop_shape_;
+  std::vector<TensorShape<>> input_shape_;
+  int ndim_ = -1;
+};
+
+template <>
+RandomBBoxCrop<CPUBackend>::~RandomBBoxCrop() = default;
+
+template <>
+RandomBBoxCrop<CPUBackend>::RandomBBoxCrop(const OpSpec &spec)
+    : Operator<CPUBackend>(spec) {
+  impl_ = std::make_unique<Impl>(spec);
```

Comment:
If this was delayed until SetupImpl, you could select appropriate dimensionality and greatly reduce the number of dimensionality checks inside - at the same time increasing speed, reducing memory footprint and eliminating a whole class of potential errors.

Reply:
done